### PR TITLE
fix(login): preserve runpodctl credentials in ~/.runpod/config.toml

### PR DESCRIPTION
**Pull request by vaskoz**
## Summary

- `flash login` was clobbering the top-level `apikey`/`apiurl` values that `runpodctl` writes to `~/.runpod/config.toml`. The cause: `save_api_key()` delegated to `runpod-python`'s `set_credentials(overwrite=True)`, which opens the file in `"w"` mode and rewrites it with only a `[default]` section.
- `save_api_key()` now does a line-based merge that updates only the `[default].api_key` field and preserves everything else — runpodctl's top-level keys, other profile sections, and any unrelated fields within `[default]`.

## Test plan

- [x] New unit tests cover: preserving runpodctl top-level keys, adding `[default]` when missing, preserving other profile sections, fresh-file creation
- [x] Existing `test_credentials.py` and `test_login.py` still pass (26/26)
- [x] Smoke-tested merger against real-world TOML shapes (file with both formats coexisting, only top-level keys, empty file, `[default]` without `api_key`, blank lines between sections)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/runpod_flash/core/credentials.py [hunk 1]
```diff
@@ -59,14 +71,62 @@ def save_api_key(api_key: str) -> Path:
...
         pass
     return path
 
 
+def _toml_quote(value: str) -> str:
+    escaped = value.replace("\\", "\\\\").replace('"', '\\"')
+    return f'"{escaped}"'
+
+
+def _upsert_default_api_key(content: str, api_key: str) -> str:
```
Comment (deanq) [inline src/runpod_flash/core/credentials.py]: Consider replacing the manual line-based merge with `tomlkit`. The goal here — preserve foreign top-level keys, other sections, and original formatting — is exactly what `tomlkit` provides via round-trip. It would collapse `_upsert_default_api_key`, `_toml_quote`, and the three regexes into ~5 lines and eliminate the inline-comment, CRLF, escaping, and section-ordering edge cases flagged below. If there's a deliberate reason to avoid the dependency (it's already pulled in transitively by `runpod-python`), worth calling out in the docstring.

**Review by deanq — COMMENTED** [review]
Inline review — see comments. Bottom line: the manual TOML merge has several edge cases (inline comments on `[default]`, CRLF, untested escaping branches) that a `tomlkit` round-trip would eliminate. Falls back to specific edge-case requests if the manual approach stays.

File: src/runpod_flash/core/credentials.py [hunk 2]
```diff
@@ -8,18 +8,26 @@
...
-    set_credentials,
 )
 
 log = logging.getLogger(__name__)
 
+# runpodctl writes top-level `apikey`/`apiurl` keys into the same config.toml
+# that runpod-python uses for its `[default]` profile. We must preserve those
+# (and any other unrelated content) when updating flash's api_key, so flash
+# login does not clobber runpodctl's credentials.
+_DEFAULT_HEADER_RE = re.compile(r"^\s*\[default\]\s*$")
```
Comment (deanq) [inline src/runpod_flash/core/credentials.py]: `^\s*\[default\]\s*$` rejects valid TOML like `[default] # flash profile`. When that's present the function falls through to the "append new section" branch and writes a second `[default]` block, producing a file `tomllib` will refuse to load on the next `flash login`. Either tighten the regex to allow trailing comments or switch to a TOML parser.

File: src/runpod_flash/core/credentials.py [hunk 3]
```diff
@@ -59,14 +71,62 @@ def save_api_key(api_key: str) -> Path:
...
+    path.write_text(new_content, encoding="utf-8")
+
     try:
         os.chmod(path, 0o600)
     except OSError:
         pass
     return path
 
 
+def _toml_quote(value: str) -> str:
```
Comment (deanq) [inline src/runpod_flash/core/credentials.py]: `_toml_quote` handles `\` and `"` escaping but no test exercises either branch. A regression that drops the backslash replacement would only surface on a pathological key. One round-trip test through `tomllib` with a key containing both characters would lock the contract.

File: src/runpod_flash/core/credentials.py [hunk 4]
```diff
@@ -59,14 +71,62 @@ def save_api_key(api_key: str) -> Path:
...
+            break
+
+    if default_start is None:
+        suffix = "" if content.endswith("\n") else "\n"
+        separator = "\n" if content.strip() else ""
+        return f"{content}{suffix}{separator}[default]\n{new_line}\n"
+
+    for i in range(default_start + 1, default_end):
+        if _API_KEY_LINE_RE.match(lines[i]):
+            ending = "\n" if lines[i].endswith("\n") else ""
```
Comment (deanq) [inline src/runpod_flash/core/credentials.py]: CRLF handling: `splitlines(keepends=True)` preserves `\r\n`, but this replacement uses `ending = "\n" if lines[i].endswith("\n") else ""`, which strips the `\r`. A file edited on Windows ends up with mixed line endings after one `flash login`. Either preserve the original line ending or normalize on read.

File: src/runpod_flash/core/credentials.py [hunk 5]
```diff
@@ -59,14 +71,62 @@ def save_api_key(api_key: str) -> Path:
...
+    escaped = value.replace("\\", "\\\\").replace('"', '\\"')
+    return f'"{escaped}"'
+
+
+def _upsert_default_api_key(content: str, api_key: str) -> str:
+    """Update `[default].api_key` in TOML text, leaving the rest intact."""
+    new_line = f"api_key = {_toml_quote(api_key)}"
+
+    if not content:
+        return f"[default]\n{new_line}\n"
```
Comment (deanq) [inline src/runpod_flash/core/credentials.py]: Nit: the literal `"default"` / `"[default]"` appears in the regex (line 27), this empty-file template, and the append template below. Per the project's "magic values as named constants" guideline, consider extracting `_DEFAULT_SECTION = "default"` and deriving the rest.

File: tests/unit/test_credentials.py [hunk 6]
```diff
@@ -69,3 +75,56 @@ def test_sets_restrictive_permissions(self, isolate_credentials_file):
...
+        assert parsed["default"]["api_key"] == "new-default"
+        assert parsed["staging"]["api_key"] == "staging-key"
+        assert parsed["staging"]["extra"] == "preserved"
+
+    def test_creates_file_with_only_default_when_missing(
+        self, isolate_credentials_file
+    ):
+        save_api_key("first-key")
+        parsed = tomllib.loads(isolate_credentials_file.read_text())
+        assert parsed == {"default": {"api_key": "first-key"}}
```
Comment (deanq) [inline tests/unit/test_credentials.py:128]: Missing unit test: the PR description mentions smoke-testing `[default]` without an `api_key` field, but there's no unit test for it. This is the only path that exercises the `insert_idx` walk-back over trailing blank lines in `_upsert_default_api_key`. Suggested:

```python
def test_inserts_api_key_into_default_without_existing_key(self, isolate_credentials_file):
    isolate_credentials_file.parent.mkdir(parents=True, exist_ok=True)
    isolate_credentials_file.write_text('[default]\nfoo = "bar"\n\n[other]\nx = 1\n')
    save_api_key("new-key")
    parsed = tomllib.loads(isolate_credentials_file.read_text())
    assert parsed["default"]["api_key"] == "new-key"
    assert parsed["default"]["foo"] == "bar"
    assert parsed["other"]["x"] == 1
```

File: tests/unit/test_credentials.py [hunk 7]
```diff
@@ -69,3 +75,56 @@ def test_sets_restrictive_permissions(self, isolate_credentials_file):
...
+        isolate_credentials_file.write_text(
+            "apikey = 'rpa_runpodctl_key'\n"
+            "apiurl = 'https://api.runpod.io/graphql'\n"
+            "\n"
+            "[default]\n"
+            'api_key = "old-flash-key"\n'
+        )
+        save_api_key("new-flash-key")
+        text = isolate_credentials_file.read_text()
+        assert "apikey = 'rpa_runpodctl_key'" in text
```
Comment (deanq) [inline tests/unit/test_credentials.py]: Nit: these substring assertions are brittle to whitespace changes and duplicate the contract already verified by the `tomllib.loads` block below. Dropping them keeps the test behavior-focused.

**Review by runpod-Henrik — COMMENTED** [review]
# Henrik's AI-Powered Bug Finder — PR #333 Review

**Verdict: NEEDS WORK**

The bug being fixed is real — `set_credentials(overwrite=True)` from runpod-python opens the file in `"w"` mode and obliterates runpodctl's top-level `apikey`/`apiurl` keys. The fix is in the right direction. CI green on 3.10/3.11/3.12.

The implementation reaches for hand-rolled TOML parsing via regex, and the section-boundary detection has at least one edge case that **silently destroys an unrelated profile's credentials**. Worth addressing before merge.

---

### 1. Bug: a `[default] # comment` or `[staging] # comment` section header makes the merger overwrite the wrong section's `api_key`

User scenario: a user has a config file someone (or another tool) has annotated with inline comments on the section headers — common in shared dev environments and pasted snippets:

\`\`\`toml
[default]
some_other_field = "x"

[staging]  # production environment
api_key = "staging-key-here"
\`\`\`

User runs `flash login`. Expected: `[default].api_key` is set to the new key, `[staging]` untouched. Actual: `[staging].api_key` is overwritten with the new flash key, and `[default]` never gets an `api_key` line.

Walk through `_upsert_default_api_key`:

1. `_DEFAULT_HEADER_RE.match` on `[default]` → matches at line 0. `default_start = 0`.
2. Inner loop scans for the next section header:
   - line 1 (`some_other_field = "x"`) → no match
   - line 2 (`[staging]  # production environment`) — `_SECTION_HEADER_RE = r"^\s*\[[^\]]+\]\s*$"` requires the line to *end* after the closing bracket and optional whitespace. The inline comment breaks the match. → **no match**
   - line 3 (`api_key = "staging-key-here"`) → no match
3. Inner loop exits without finding a section boundary. `default_end = len(lines)`.
4. Replacement scan from `default_start+1` to `default_end`: line 3 matches `_API_KEY_LINE_RE` → gets replaced with `api_key = "<new flash key>"`.
5. The user's staging credentials are silently destroyed. No warning, no log.

Repro test (suggest adding):

\`\`\`python
def test_inline_comment_on_section_header_does_not_overwrite_other_profile(
    self, isolate_credentials_file
):
    isolate_credentials_file.parent.mkdir(parents=True, exist_ok=True)
    isolate_credentials_file.write_text(
        "[default]\n"
        "some_other_field = \"x\"\n"
        "\n"
        "[staging]  # production environment\n"
        'api_key = "staging-key"\n'
    )
    save_api_key("new-flash-key")
    parsed = tomllib.loads(isolate_credentials_file.read_text())
    assert parsed["default"]["api_key"] == "new-flash-key"
    assert parsed["staging"]["api_key"] == "staging-key"   # currently FAILS
\`\`\`

The cleanest fix is the one another reviewer already suggested: use `tomlkit` (round-trips comments + structure), or use stdlib `tomllib`/`tomli` to parse + Python serializer to write. The hand-rolled regex approach will keep producing edge cases like this.

If staying with regex for some reason: tighten the section-header regex to accept inline comments, e.g.

\`\`\`python
_SECTION_HEADER_RE = re.compile(r"^\s*\[[^\]]+\]\s*(#.*)?$")
_DEFAULT_HEADER_RE = re.compile(r"^\s*\[default\]\s*(#.*)?$")
\`\`\`

And add a test for `[default] # comment` too.

---

### 2. Issue: incomplete TOML string escaping

`_toml_quote` escapes only `\` and `"`. The TOML spec requires escaping control characters (U+0000–U+001F and U+007F) inside basic strings. User scenario: a future deploy or auth path passes an `api_key` containing a tab or newline (unlikely for prod API keys, but plausible for trimmed-from-clipboard edge cases — and PR #337 ships a new copy-paste flow where the user could paste extra whitespace). The written file becomes invalid TOML, and the next `flash login` or `flash deploy` fails at credential load with an opaque parser error.

Two cheap mitigations:
- Use TOML literal strings (`'...'` single quotes — no escape sequences allowed, but no control chars either). Safer for keys that are alphanumeric anyway.
- Reject the api_key with a clear error if it contains non-printable chars before writing.

Or — again — let `tomlkit`/`tomllib` handle serialization.

---

### 3. Issue: line endings drift on Windows / CRLF files

`content.splitlines(keepends=True)` preserves the original `\r\n` for existing lines, but the new `api_key = ...` line is appended with only `\n`. A Windows-edited config will get a `\n`-terminated line in a sea of `\r\n` lines after running `flash login`.

Most editors and tools handle mixed line endings, so this is cosmetic. But it'll show up as a single-line diff in version control or in editors that surface line-ending mismatches. Worth detecting the predominant ending and matching it, or normalizing on write.

---

### 4. Question: behavior when `[default]` exists with no `api_key` and the section has trailing blank lines

The insert path:

\`\`\`python
insert_idx = default_end
while insert_idx > default_start + 1 and lines[insert_idx - 1].strip() == "":
    insert_idx -= 1
lines.insert(insert_idx, new_line + "\n")
\`\`\`

This walks back over blank lines and inserts just below the last non-blank line in `[default]`. Looks correct, but no test covers it. Worth one:

\`\`\`python
def test_inserts_api_key_above_trailing_blank_lines_in_default(...):
    text = "[default]\nname = \"primary\"\n\n\n[staging]\n..."
    # After: api_key sits next to name, blank lines preserved between [default] and [staging]
\`\`\`

Not a bug claim — just untested behavior worth pinning down.

---

### 5. Nit: line-based regex on TOML doesn't survive multi-line strings

If `[default]` ever contains a multi-line basic string (`description = """\n...\n"""`) whose content happens to include a line matching `^\s*api_key\s*=`, the merger would replace inside the string literal. Vanishingly unlikely for the credentials file but worth noting that a real TOML parser would eliminate the class of edge cases entirely. Same recommendation as #1.

---

### Nits

- **Test coverage is good for the documented cases** — preserve top-level keys, missing `[default]`, other profile sections, fresh file. Just missing the comment-on-section-header case (above) and the trailing-blank-lines case (above).
- **The `_OLD_XDG_PATH` migration path is untouched** — separate concern, but worth confirming the migration helper still works correctly when the source file has the same multi-format shape (runpodctl top-level + flash `[default]`).
- **0600 perms set after write** — good. Worth confirming this works on Windows (where `os.chmod` is a partial no-op); current `try/except OSError` swallows it correctly, so Windows users just won't have restrictive perms, which is the existing behavior.

---

🤖 *Reviewed by Henrik's AI-Powered Bug Finder*

**Comment by deanq** [comment]
Pushed `dfbc9ca` addressing the review feedback by replacing the regex merge with a `tomlkit` round-trip — the approach Copilot, and Henrik's bug-finder all converged on. This resolves the edge cases structurally rather than patching them one regex at a time.

**Implementation**
- `_upsert_default_api_key` now parses with `tomlkit`, sets `[default].api_key`, and dumps. Foreign top-level keys (runpodctl's `apikey`/`apiurl`), sibling profiles, comments, and escaping are preserved by construction.
- `save_api_key` reads/writes with `newline=""` so CRLF files keep their endings (`read_text` was collapsing `\r\n` → `\n`).
- Dropped the three header/key regexes and `_toml_quote`; added `_DEFAULT_SECTION`/`_API_KEY_FIELD` constants.
- Declared `tomlkit` as a direct dependency (was transitive via `runpod`).

**How each item is covered (all with regression tests)**
- Inline-comment section header overwriting another profile → fixed; `[staging]` left untouched.
- `[default] # comment` → updated in place, no duplicate table.
- `[default]` with no trailing newline → valid TOML output.
- Backslash / double-quote / control-char keys → round-trip through `tomllib`.
- CRLF preservation → byte-level assertion.
- `[default]` without an `api_key` field → key inserted, siblings preserved.
- Dropped the brittle substring assertions (kept the `tomllib`-parsed checks).

**One behavior change worth noting:** `tomlkit.parse` raises on malformed existing TOML, where the old regex appended regardless. I kept it fail-loud rather than reintroduce clobber-on-corrupt, since silently overwriting is exactly the data loss this PR fixes. Happy to add a guarded fallback if reviewers prefer.

`make quality-check` green: 2574 + 53 tests pass, coverage 85.94%.

File: src/runpod_flash/core/credentials.py [hunk 8]
```diff
@@ -59,14 +72,46 @@ def save_api_key(api_key: str) -> Path:
...
+def _upsert_default_api_key(content: str, api_key: str) -> str:
+    """Update `[default].api_key` in TOML text, leaving the rest intact.
+
+    Parses with tomlkit so foreign top-level keys (runpodctl's
+    `apikey`/`apiurl`), sibling profile sections, comments, and the file's
+    original line endings are preserved. Only `[default].api_key` is mutated;
+    a missing `[default]` table is created.
+    """
+    doc = tomlkit.parse(content) if content else tomlkit.document()
+
```
[the remark this answers is not in the document]
Comment (deanq) [reply src/runpod_flash/core/credentials.py]: Done in a17a065. `_upsert_default_api_key` now catches `tomlkit.exceptions.TOMLKitError`, logs a warning, and falls back to a fresh `[default]` document so `flash login` completes. A malformed file is already unloadable (runpodctl can't read it either), so there are no recoverable credentials to preserve. Added `test_recovers_from_corrupt_existing_file` covering it.

**Comment by deanq** [comment]
CI is green and every review thread is resolved — this is ready for another look.

The earlier red on `Quality Gates (3.12)` was unrelated to this change: `test_regressions.py::TestREG008` failing with `_pickle.PicklingError`. It passed on 3.10/3.11/3.13 and reruns clean. Root-caused separately as a background-thread state leak in `deploy_all_background`, fixed in #361 — nothing to do with the credentials merge here.

@runpod-Henrik re-requesting review — the `tomlkit` round-trip in `f247469` plus the malformed-TOML fallback in `10a2b56` cover every item from your bug-finder pass, each with a regression test.

**Review by deanq — APPROVED** [review]
